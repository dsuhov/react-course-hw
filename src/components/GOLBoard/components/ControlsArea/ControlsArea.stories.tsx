import React from "react";
import { ControlsArea } from "./ControlsArea";
export default {
  title: "Controls Area",
};

import { Provider } from "react-redux";
import { store } from "@/rdx/store";

export const controlsAreaBoardRunning = () => {
  return (
    <Provider store={store}>
      <ControlsArea />
    </Provider>
  );
};
