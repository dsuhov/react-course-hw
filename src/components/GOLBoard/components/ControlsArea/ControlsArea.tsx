import React from "react";
import {
  ControlsWrapper,
  ControlsLine,
  Buttons,
  FieldSize,
  Label,
  StartBlock,
  NumberInput,
  Button,
} from "./styled";

interface ControlsAreaProps {
  status: "running" | "paused" | "stopped";
  cmdBtnHadler: (evt: React.MouseEvent) => void;
  cmdFormHandler: (params: {
    size: [number, number];
    fullness: number;
  }) => void;
}

export class ControlsArea extends React.Component<ControlsAreaProps, {}> {
  xSizeInput = React.createRef<HTMLInputElement>();
  ySizeInput = React.createRef<HTMLInputElement>();
  fullnessInput = React.createRef<HTMLInputElement>();

  shouldComponentUpdate(nextProps: ControlsAreaProps) {
    return this.props.status !== nextProps.status;
  }

  handleSubmit = (evt: React.FormEvent) => {
    evt.preventDefault();

    const sizeX = this.xSizeInput.current!.value;
    const sizeY = this.ySizeInput.current!.value;
    const fullness = this.fullnessInput.current!.value;

    this.props.cmdFormHandler({
      size: [+sizeX, +sizeY],
      fullness: +fullness,
    });
  };

  render() {
    const status = this.props.status;
    const cmdBtnHadler = this.props.cmdBtnHadler;

    return (
      <ControlsWrapper>
        <ControlsLine>
          <Buttons>
            <Button
              name="faster"
              onClick={cmdBtnHadler}
              disabled={status !== "running"}
            >
              Faster
            </Button>
            <Button
              name="slower"
              onClick={cmdBtnHadler}
              disabled={status !== "running"}
            >
              Slower
            </Button>
            <Button
              name={status === "paused" ? "resume" : "pause"}
              onClick={cmdBtnHadler}
              disabled={false}
            >
              {status === "paused" ? "Resume" : "Pause"}
            </Button>
            <Button
              name="reset"
              onClick={cmdBtnHadler}
              disabled={status !== "paused"}
            >
              Reset
            </Button>
          </Buttons>
        </ControlsLine>

        <form onSubmit={this.handleSubmit}>
          <ControlsLine>
            <FieldSize>
              <Label>Field Size:</Label>
              <NumberInput
                type="number"
                min={2}
                max={200}
                name="xSize"
                ref={this.xSizeInput}
                required
              />
              <span> x </span>
              <NumberInput
                type="number"
                min={2}
                max={200}
                name="ySize"
                ref={this.ySizeInput}
                required
              />
            </FieldSize>
            <div>
              <Label>Fullness:</Label>
              <NumberInput
                type="number"
                name="fullness"
                min={10}
                max={100}
                ref={this.fullnessInput}
                required
              />
              <span> %</span>
            </div>
          </ControlsLine>

          <ControlsLine>
            <StartBlock>
              <button
                type="submit"
                name="start"
                disabled={status === "running" || status === "paused"}
              >
                Start
              </button>
            </StartBlock>
          </ControlsLine>
        </form>
      </ControlsWrapper>
    );
  }
}
