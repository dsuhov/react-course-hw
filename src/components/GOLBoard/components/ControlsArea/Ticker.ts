export class Ticker {
  private trigger: () => void;
  private intId = 0;
  private interval = 200;

  constructor(trigger: () => void) {
    this.trigger = trigger;
  }

  public start() {
    if (this.intId) {
      window.clearInterval(this.intId);
    }

    this.intId = window.setInterval(() => {
      this.trigger();
    }, this.interval);
  }

  public stop() {
    if (!this.intId) {
      return;
    }

    window.clearInterval(this.intId);
    this.intId = 0;
  }

  public getSpeed() {
    return this.interval;
  }

  public setSpeed(interval: number) {
    this.interval = interval;
  }

  public setFaster() {
    this.stop();
    const newInt = this.interval - 100;

    if (newInt < 200) {
      this.start();
      return;
    }
    this.interval = newInt;
    this.start();
  }

  public setSlower() {
    this.stop();
    this.interval = this.interval + 200;
    this.start();
  }
}
