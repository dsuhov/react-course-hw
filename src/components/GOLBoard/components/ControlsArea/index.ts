export * from "./ControlsArea";
export * from "./saga";
