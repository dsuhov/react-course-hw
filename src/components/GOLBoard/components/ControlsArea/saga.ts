import { put, takeEvery } from "redux-saga/effects";
import { gameFieldActions } from "@/rdx/gameField/gameFieldSlice";
import { gameStatusActions } from "@/rdx/gameStatus/gameStatusSlice";

export function* incGeneration() {
  yield put(gameStatusActions.incGen());
}

export function* watchUpdateField() {
  yield takeEvery(gameFieldActions.updateField.type, incGeneration);
}
