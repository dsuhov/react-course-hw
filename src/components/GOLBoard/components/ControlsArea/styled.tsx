import styled from "@emotion/styled";
import { css } from "@emotion/core";

export const ControlsWrapper = styled.div`
  padding-top: 14px;
`;

export const ControlsLine = styled.div`
  display: flex;
  justify-content: center;
  margin-bottom: 10px;
`;

export const Buttons = styled.div`
  display: flex;

  & > *:not(:last-child) {
    margin-right: 8px;
  }
`;

export const FieldSize = styled.div`
  margin-right: 20px;
`;

export const Label = styled.div`
  margin-bottom: 6px;
`;

export const StartBlock = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;

  button {
    padding: 8px 12px;
    font-size: 1.1em;
    display: block;
    text-transform: uppercase;
    cursor: pointer;
    border: 2px solid #9eb7e5;
    border-radius: 4px;
    color: #304c89;
    background: #f2efe4;
    font-weight: 700;
    letter-spacing: 0.06em;

    &:hover {
      box-shadow: 0 0 6px #9eb7e5;
      background: #f6f5ea;
    }
  }
`;

export const NumberInput = styled.input`
  padding: 5px 10px;
  border: 2px solid #9eb7e5;
  border-radius: 4px;
  max-width: 80px;

  &:focus {
    outline: 1px solid #648de5;
    box-shadow: 0 0 5px #648de5;
  }
`;

export const Button = styled.button<{ disabled: boolean }>`
  display: block;
  text-transform: uppercase;
  cursor: pointer;
  padding: 5px 10px;
  border: 2px solid #9eb7e5;
  border-radius: 4px;
  color: #304c89;
  background: #f2efe4;
  font-weight: 700;
  letter-spacing: 0.06em;

  &:hover {
    box-shadow: 0 0 6px #9eb7e5;
    background: #f6f5ea;
  }

  ${({ disabled }) => {
    if (disabled) {
      return css`
        opacity: 0.6;
        pointer-events: none;
      `;
    }
  }}
`;
