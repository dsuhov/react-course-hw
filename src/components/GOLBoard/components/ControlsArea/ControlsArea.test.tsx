import React from "react";
import { mount } from "enzyme";
import { ControlsArea } from "./ControlsArea";
import renderer from "react-test-renderer";

describe("<ControlsArea /> tests", () => {
  it("Snapshot for <ControlsArea />", () => {
    expect(
      renderer
        .create(
          <ControlsArea
            status="running"
            cmdBtnHadler={jest.fn()}
            cmdFormHandler={jest.fn()}
          />
        )
        .toJSON()
    ).toMatchSnapshot();
  });

  it("Check control buttons functionality, game running", () => {
    const btns = ["faster", "slower", "pause"];
    const cmdBtnHadler = jest.fn((evt) => evt.target.name);

    const area = mount(
      <ControlsArea
        status="running"
        cmdBtnHadler={cmdBtnHadler}
        cmdFormHandler={jest.fn()}
      />
    );

    btns.forEach((btn) => {
      area.find(`button[name="${btn}"]`).simulate("click");
    });

    expect(cmdBtnHadler.mock.calls.length).toBe(btns.length);

    btns.forEach((btn, i) => {
      expect(cmdBtnHadler.mock.results[i].value).toBe(btn);
    });

    expect(area.find(`button[name="${"reset"}"]`).getDOMNode()).toHaveProperty(
      "disabled"
    );
  });

  it("Check control buttons functionality, game paused", () => {
    const btns = ["faster", "slower", "resume"];
    const cmdBtnHadler = jest.fn((evt) => evt.target.name);

    const area = mount(
      <ControlsArea
        status="paused"
        cmdBtnHadler={cmdBtnHadler}
        cmdFormHandler={jest.fn()}
      />
    );

    btns.forEach((btn) => {
      expect(area.find(`button[name="${btn}"]`).getDOMNode()).toHaveProperty(
        "disabled"
      );
    });

    area.find(`button[name="reset"]`).simulate("click");
    expect(cmdBtnHadler.mock.results[0].value).toBe("reset");
  });

  it("Check control buttons functionality, game stopped", () => {
    const btns = ["faster", "slower", "pause", "reset"];

    const area = mount(
      <ControlsArea
        status="stopped"
        cmdBtnHadler={jest.fn()}
        cmdFormHandler={jest.fn()}
      />
    );

    btns.forEach((btn) => {
      expect(area.find(`button[name="${btn}"]`).getDOMNode()).toHaveProperty(
        "disabled"
      );
    });
  });
});
