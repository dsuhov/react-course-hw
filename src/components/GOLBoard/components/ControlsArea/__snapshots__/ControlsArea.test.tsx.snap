// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ControlsArea /> tests Snapshot for <ControlsArea /> 1`] = `
.emotion-16 {
  padding-top: 14px;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-bottom: 10px;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-4 > *:not(:last-child) {
  margin-right: 8px;
}

.emotion-0 {
  display: block;
  text-transform: uppercase;
  cursor: pointer;
  padding: 5px 10px;
  border: 2px solid #9eb7e5;
  border-radius: 4px;
  color: #304c89;
  background: #f2efe4;
  font-weight: 700;
  -webkit-letter-spacing: 0.06em;
  -moz-letter-spacing: 0.06em;
  -ms-letter-spacing: 0.06em;
  letter-spacing: 0.06em;
  opacity: 0.6;
  pointer-events: none;
}

.emotion-0:hover {
  box-shadow: 0 0 6px #9eb7e5;
  background: #f6f5ea;
}

.emotion-9 {
  margin-right: 20px;
}

.emotion-6 {
  margin-bottom: 6px;
}

.emotion-7 {
  padding: 5px 10px;
  border: 2px solid #9eb7e5;
  border-radius: 4px;
  max-width: 80px;
}

.emotion-7:focus {
  outline: 1px solid #648de5;
  box-shadow: 0 0 5px #648de5;
}

.emotion-14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-14 button {
  padding: 8px 12px;
  font-size: 1.1em;
  display: block;
  text-transform: uppercase;
  cursor: pointer;
  border: 2px solid #9eb7e5;
  border-radius: 4px;
  color: #304c89;
  background: #f2efe4;
  font-weight: 700;
  -webkit-letter-spacing: 0.06em;
  -moz-letter-spacing: 0.06em;
  -ms-letter-spacing: 0.06em;
  letter-spacing: 0.06em;
}

.emotion-14 button:hover {
  box-shadow: 0 0 6px #9eb7e5;
  background: #f6f5ea;
}

.emotion-13 {
  display: block;
  text-transform: uppercase;
  cursor: pointer;
  padding: 5px 10px;
  border: 2px solid #9eb7e5;
  border-radius: 4px;
  color: #304c89;
  background: #f2efe4;
  font-weight: 700;
  -webkit-letter-spacing: 0.06em;
  -moz-letter-spacing: 0.06em;
  -ms-letter-spacing: 0.06em;
  letter-spacing: 0.06em;
}

.emotion-13:hover {
  box-shadow: 0 0 6px #9eb7e5;
  background: #f6f5ea;
}

<div
  className="emotion-16"
>
  <div
    className="emotion-5"
  >
    <div
      className="emotion-4"
    >
      <button
        className="emotion-0"
        disabled={true}
        name="faster"
        onClick={[Function]}
      >
        Faster
      </button>
      <button
        className="emotion-0"
        disabled={true}
        name="slower"
        onClick={[Function]}
      >
        Slower
      </button>
      <button
        className="emotion-0"
        disabled={true}
        name="pause"
        onClick={[Function]}
      >
        Pause
      </button>
      <button
        className="emotion-0"
        disabled={true}
        name="reset"
        onClick={[Function]}
      >
        Reset
      </button>
    </div>
  </div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="emotion-5"
    >
      <div
        className="emotion-9"
      >
        <div
          className="emotion-6"
        >
          Field Size:
        </div>
        <input
          className="emotion-7"
          max={200}
          min={2}
          name="xSize"
          required={true}
          type="number"
        />
        <span>
           x 
        </span>
        <input
          className="emotion-7"
          max={200}
          min={2}
          name="ySize"
          required={true}
          type="number"
        />
      </div>
      <div>
        <div
          className="emotion-6"
        >
          Fullness:
        </div>
        <input
          className="emotion-7"
          max={100}
          min={10}
          name="fullness"
          required={true}
          type="number"
        />
        <span>
           %
        </span>
      </div>
    </div>
    <div
      className="emotion-5"
    >
      <div
        className="emotion-14"
      >
        <button
          className="emotion-13"
          disabled={false}
          name="start"
          type="submit"
        >
          Start
        </button>
      </div>
    </div>
  </form>
</div>
`;
